File: frontend/src/app/services/support.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { SupportTicket, SupportRespondRequest } from '../models/support-ticket.model';
import { environment } from '../../environments/environment';

/**
 * Service for managing escalated support tickets.
 */
@Injectable({
  providedIn: 'root'
})
export class SupportService {
  private readonly apiUrl = `${environment.apiUrl}/support`;

  constructor(private http: HttpClient) {}

  /**
   * Retrieves all support tickets, optionally filtered by status.
   */
  getAllTickets(status?: string): Observable<SupportTicket[]> {
    let params = new HttpParams();
    if (status) {
      params = params.set('status', status);
    }
    return this.http.get<SupportTicket[]>(`${this.apiUrl}/tickets`, { params });
  }

  /**
   * Retrieves a single ticket by ID.
   */
  getTicketById(id: number): Observable<SupportTicket> {
    return this.http.get<SupportTicket>(`${this.apiUrl}/tickets/${id}`);
  }

  /**
   * Adds a human agent's response to a ticket.
   */
  respondToTicket(id: number, response: string): Observable<SupportTicket> {
    const request: SupportRespondRequest = { response };
    return this.http.post<SupportTicket>(`${this.apiUrl}/respond/${id}`, request);
  }

  /**
   * Marks a ticket as resolved.
   */
  resolveTicket(id: number): Observable<SupportTicket> {
    return this.http.put<SupportTicket>(`${this.apiUrl}/resolve/${id}`, {});
  }

  /**
   * Returns a CSS class for the ticket status badge.
   */
  getStatusClass(status: string): string {
    switch (status) {
      case 'OPEN': return 'status-open';
      case 'IN_PROGRESS': return 'status-in-progress';
      case 'RESOLVED': return 'status-resolved';
      default: return '';
    }
  }

  /**
   * Returns a display-friendly label for ticket status.
   */
  getStatusLabel(status: string): string {
    switch (status) {
      case 'OPEN': return 'Open';
      case 'IN_PROGRESS': return 'In Progress';
      case 'RESOLVED': return 'Resolved';
      default: return status;
    }
  }
}

File: backend/backend/src/main/java/com/artemisdocs/backend/service/SupportService.java
package com.artemisdocs.backend.service;

import com.artemisdocs.backend.dto.SupportTicketResponse;
import com.artemisdocs.backend.entity.Document;
import com.artemisdocs.backend.entity.SupportTicket;
import com.artemisdocs.backend.entity.SupportTicket.TicketStatus;
import com.artemisdocs.backend.exception.GlobalExceptionHandler.EntityNotFoundException;
import com.artemisdocs.backend.repository.DocumentRepository;
import com.artemisdocs.backend.repository.SupportTicketRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;

import java.time.LocalDateTime;
import java.util.List;

/**
 * Service for managing escalated support tickets.
 * Provides CRUD operations and status lifecycle management.
 */
@Service
@Slf4j
@RequiredArgsConstructor
public class SupportService {

    private final SupportTicketRepository supportTicketRepository;
    private final DocumentRepository documentRepository;

    /**
     * Retrieves all support tickets, optionally filtered by status.
     *
     * @param status optional status filter
     * @return list of ticket response DTOs
     */
    public List<SupportTicketResponse> getAllTickets(String status) {
        List<SupportTicket> tickets;

        if (status != null && !status.isBlank()) {
            try {
                TicketStatus ticketStatus = TicketStatus.valueOf(status.toUpperCase());
                tickets = supportTicketRepository.findByStatus(ticketStatus);
            } catch (IllegalArgumentException e) {
                throw new IllegalArgumentException("Invalid status: " + status +
                        ". Valid values are: OPEN, IN_PROGRESS, RESOLVED");
            }
        } else {
            tickets = supportTicketRepository.findAllByOrderByCreatedAtDesc();
        }

        return tickets.stream()
                .map(this::toResponse)
                .toList();
    }

    /**
     * Retrieves a single ticket by ID.
     *
     * @param id the ticket ID
     * @return ticket response DTO
     */
    public SupportTicketResponse getTicketById(Long id) {
        SupportTicket ticket = findTicketOrThrow(id);
        return toResponse(ticket);
    }

    /**
     * Adds a human agent's response to a ticket and marks it IN_PROGRESS.
     *
     * @param id       the ticket ID
     * @param response the human agent's response text
     * @return updated ticket response DTO
     */
    public SupportTicketResponse respondToTicket(Long id, String response) {
        SupportTicket ticket = findTicketOrThrow(id);

        ticket.setResponse(response);
        ticket.setStatus(TicketStatus.IN_PROGRESS);

        ticket = supportTicketRepository.save(ticket);
        log.info("Response added to ticket {}", id);

        return toResponse(ticket);
    }

    /**
     * Marks a ticket as resolved.
     *
     * @param id the ticket ID
     * @return updated ticket response DTO
     */
    public SupportTicketResponse resolveTicket(Long id) {
        SupportTicket ticket = findTicketOrThrow(id);

        ticket.setStatus(TicketStatus.RESOLVED);
        ticket.setResolvedAt(LocalDateTime.now());

        ticket = supportTicketRepository.save(ticket);
        log.info("Ticket {} resolved", id);

        return toResponse(ticket);
    }

    /**
     * Finds a ticket by ID or throws EntityNotFoundException.
     */
    private SupportTicket findTicketOrThrow(Long id) {
        return supportTicketRepository.findById(id)
                .orElseThrow(() -> new EntityNotFoundException("Support ticket not found with ID: " + id));
    }

    /**
     * Converts a SupportTicket entity to its response DTO,
     * enriching it with the associated document filename.
     */
    private SupportTicketResponse toResponse(SupportTicket ticket) {
        String documentFilename = null;
        if (ticket.getDocumentId() != null) {
            documentFilename = documentRepository.findById(ticket.getDocumentId())
                    .map(Document::getFilename)
                    .orElse("Unknown Document");
        }

        return new SupportTicketResponse(
                ticket.getId(),
                ticket.getSessionId(),
                ticket.getQuestion(),
                ticket.getContext(),
                ticket.getStatus().name(),
                ticket.getResponse(),
                ticket.getAiAnswer(),
                ticket.getConfidenceScore(),
                documentFilename,
                ticket.getDocumentId(),
                ticket.getCreatedAt(),
                ticket.getResolvedAt()
        );
    }
}
